AK/String.cpp

```diff
@@ -0,0 +1,338 @@
+    }
+
+    StringView bytes_as_string_view() const { return { bytes() }; }
+
+    bool operator==(StringData const& other) const
+    {
+        return bytes_as_string_view() == other.bytes_as_string_view();
+    }
+
+    unsigned hash() const
+    {
+        if (!m_has_hash)
+            compute_hash();
+        return m_hash;
+    }
+
+private:
+    explicit StringData(size_t byte_count);
+    StringData(StringData const& superstring, size_t start, size_t byte_count);
+
+    void compute_hash() const;
+
+    u32 m_byte_count { 0 };
+    mutable unsigned m_hash { 0 };
+    mutable bool m_has_hash { false };
+    bool m_substring { false };
+
+    u8 m_bytes_or_substring_data[0];
+};
+
+String::String(NonnullRefPtr<StringData> data)
+    : m_data(&data.leak_ref())
+{
+}
+
+String::String(ShortString short_string)
+    : m_short_string(short_string)
+{
+}
+
+String::String(String const& other)
+    : m_data(other.m_data)
+{
+    if (!is_short_string())
+        m_data->ref();
+}
+
+String::String(String&& other)
+    : m_data(exchange(other.m_data, nullptr))
+{
+}
+
+String& String::operator=(String&& other)
+{
+    m_data = exchange(other.m_data, nullptr);
+    return *this;
+}
+
+String& String::operator=(String const& other)
+{
+    if (&other != this) {
+        m_data = other.m_data;
+        if (!is_short_string())
+            m_data->ref();
+    }
+    return *this;
+}
+
+String::~String()
+{
+    if (!is_short_string() && m_data)
+        m_data->unref();
+}
+
+String::String()
+{
+    m_short_string.byte_count_and_type = 0x01;
+}
+
+ErrorOr<String> String::from_utf8(StringView view)
+{
+    if (view.length() <= MAX_SHORT_STRING_LENGTH) {
+        ShortString short_string;
+        if (!view.is_empty())
+            memcpy(short_string.storage, view.characters_without_null_termination(), view.length());
+        short_string.byte_count_and_type = (view.length() << 1) | 0x01;
+        return String { short_string };
+    }
+    auto data = TRY(StringData::from_utf8(view.characters_without_null_termination(), view.length()));
+    return String { data };
+}
+
+StringView String::bytes_as_string_view() const
+{
+    return StringView(bytes());
+}
+
+ReadonlyBytes String::bytes() const
+{
+    if (is_short_string())
+        return m_short_string.bytes();
+    return m_data->bytes();
+}
+
+bool String::is_empty() const
+{
+    return bytes().size() == 0;
+}
+
+ErrorOr<String> String::vformatted(StringView fmtstr, TypeErasedFormatParams& params)
+{
+    StringBuilder builder;
+    TRY(vformat(builder, fmtstr, params));
+    return builder.to_string();
+}
+
+bool String::operator==(String const& other) const
+{
+    if (is_short_string())
+        return m_data == other.m_data;
+    return bytes_as_string_view() == other.bytes_as_string_view();
+}
+
+bool String::operator==(StringView other) const
+{
+    return bytes_as_string_view() == other;
+}
+
+bool String::operator<(String const& other) const
```

Comment:
I'd be wary about introducing `operator<` and `operator>` as they don't work for general alphabetical sorting of unicode strings (think a/ã/ä/â/...), instead we should have (already might have?) something like `localeCompare` in JS.

Reply:
And if you think we should have it, an explicit name ("bytes lower than"?) might be a better choice.